pysindy/feature_library/base.py

```diff
@@ -256,3 +262,524 @@ def get_feature_names(self, input_features=None):
+
+    Attributes
+    ----------
+    libraries_ : list of libraries
+        Library instances to be applied to the input matrix.
+
+    tensor_array_ : 2D list of booleans, optional,
+            (default is to not tensor any of the libraries together) shape
+            equal to (# of tensor libraries to make, # feature libraries)
+        Indicates which pairs of libraries to tensor product together and
+        add to the overall library. For instance if you have 5 libraries,
+        and want to do two tensor products, you could use the list
+        [[1, 0, 0, 1, 0], [0, 1, 0, 1, 1]] to indicate that you want two
+        tensored libraries from tensoring libraries 0 and 3 and libraries
+        1, 3, and 4.
+
+    inputs_per_library_ : np.ndarray, optional
+            (default all inputs used for every library) shape equal to
+            (# feature libraries, # variable inputs)
+        Can be used to specify a subset of the variables to use to generate
+        a feature library. If number of feature libraries > 1, then can be
+        use to generate a large number of libraries, each using their own
+        subsets of the input variables. Note that this must be specified for
+        all the individual feature libraries.
+
+    n_input_features_ : int
+        The total number of input features.
+        WARNING: This is deprecated in scikit-learn version 1.0 and higher so
+        we check the sklearn.__version__ and switch to n_features_in if needed.
+
+    n_output_features_ : int
+        The total number of output features. The number of output features
+        is the sum of the numbers of output features for each of the
+        concatenated libraries.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from pysindy.feature_library import FourierLibrary, CustomLibrary
+    >>> from pysindy.feature_library import GeneralizedLibrary
+    >>> x = np.array([[0.,-1],[1.,0.],[2.,-1.]])
+    >>> functions = [lambda x : np.exp(x), lambda x,y : np.sin(x+y)]
+    >>> lib_custom = CustomLibrary(library_functions=functions)
+    >>> lib_fourier = FourierLibrary()
+    >>> lib_generalized = GeneralizedLibrary([lib_custom, lib_fourier])
+    >>> lib_generalized.fit(x)
+    >>> lib_generalized.transform(x)
+    """
+
+    def __init__(
+        self,
+        libraries: list,
+        tensor_array=None,
+        inputs_per_library=None,
+        library_ensemble=False,
+        ensemble_indices=[0],
+    ):
+        super(GeneralizedLibrary, self).__init__(
+            library_ensemble=library_ensemble, ensemble_indices=ensemble_indices
+        )
+        if len(libraries) > 0:
+            self.libraries_ = libraries
+        else:
+            raise ValueError(
+                "Empty or nonsensical library list passed to this library."
+            )
+        if inputs_per_library is not None:
+            if len(inputs_per_library.shape) != 2:
```

Comment:
Trick: numpy arrays actually have an attribute that measures this directly: `ndim`. So this line can be written as `if inputs_per_library.ndim !=2:`